docs/rfcs/2025-08-11-channels-forwarding.md

```diff
@@ -0,0 +1,188 @@
+
+**How to split forwarding jobs into smaller delivery batches:**
+
+For simplicity, this will be done only for channels, and not for inviting admins in regular groups, at least initially. Inviting admins have more complex logic, involving multiple queries and filters to retrieve less (only necessary) members, as we're trying to limit load on them, as they can be mobile clients. They will be moved to asynchronous processing too, though.
+
+Forwarding job will read required members in loop using a cursor (e.g., use group_member_id as a cursor) to split into further delivery jobs (group sends) by members. As a reminder, group send in terms of chat logic is not a network operation, but a request to agent. We can consider this part a black box for now, and consider if groups sends should be made concurrently later. For now though, for simplicity we can consider that group sends are done in a synchronous loop with moving cursor for member retrieval. Cursor position can be remembered on the job record after each group send, for recovery.
+
+It's unclear if there's a need for a separate lower level abstraction for delivery jobs, that could be reused for feeds and other purposes, as they would require different logic of building delivery lists, and scheduling a delivery for agent already serves a similar purpose. For now, it seems this additional abstraction is unnecessary.
+
+**How a forwarding job will work overall:**
+
+0. There is some process on start that launches necessary workers for each group/scope client serves as a forwarding agent. Also message receive loop "kicks" necessary workers. Below we start with a worker for some group/scope trying to retrieve next work item, that is being next message(s) to forward.
+1. Worker retrieves next message from `messages` table that was marked for forwarding, that matches the worker's group/scope.
+    - Question: Do we need a separate flag on message record to mark it as "for forwarding"? Perhaps having forward_scope (see schema below) is enough.
+2. Worker checks, if this message is already attached to some forwarding job, and the state of the job.
+    1. Worker gets job record.
+        - If message is not attached to any job yet, worker creates a new job record, sets job id on the message record (normal execution path).
+        - Otherwise worker gets existing job attached to the message (recovery path).
+    2. Worker determines forwarding batch encoding for the job.
+        - If encoding is not saved on the job, it means it wasn't determined before (normal execution):
+          1. Worker retrieves more messages that match its group/scope (in loop or in bulk? also message list has to preserve reception order).
```

Comment:
in bulk is better I think

Reply:
but we won't know how much we'd need. Some heuristic could be read 100, then mark those that are used with jobId, or load more if needed. It's important to put encoding into job and mark all relevant messages in one transaction.